File: 1070/b.py
n = int(input())

black_listed = []
white_listed = []
multiplication = [2**24, 2**16, 2**8, 1]
ranges = {2 ** i: str(32-i) for i in range(0, 33)}
for i in range(n):
    address_str = input()
    is_black_list = address_str[0] == '-'
    if '/' in address_str:
        clean_addr, x = address_str[1:].split('/')
        x = int(x)
    else:
        clean_addr, x = address_str[1:], 32
    addr_num = 0
    for multipier, num in zip(multiplication, map(int, clean_addr.split('.'))):
        addr_num += multipier * num
    address_range = pow(2, 32-x)
    if is_black_list:
        black_listed.append((addr_num, addr_num+address_range-1))
    else:
        white_listed.append((addr_num, addr_num+address_range-1))

black_listed = sorted(black_listed)
white_listed = sorted(white_listed)

ip_ranges = []


def add_ips(left, right):
    if check_list(black_listed, left, right):
        if check_list(white_listed, left, right):
            if left == right:
                print(-1)
                exit()
            boundary = (right - left) // 2
            add_ips(left, left+boundary)
            add_ips(left+boundary+1, right)
        else:
            ip_ranges.append(number_to_address(left) + '/' + ranges[right-left+1])


def check_list(list_to_check, left, right):
    if list_to_check:
        le, ri = list_to_check[find_index(list_to_check, left)]
        if left <= le <= right or left <= ri <= right or le <= left <= ri or le <= right <= ri:
            return True
    return False


def find_index(list_to_check, index):
    left, right = 0, len(list_to_check) - 1
    while True:
        if left == right:
            return left
        mid = left + ((right - left) // 2)
        if list_to_check[mid][0] <= index <= list_to_check[mid][1]:
            return mid
        if list_to_check[mid][0] > index:
            right = mid
        else:
            left = mid+1


def number_to_address(number):
    addr_strs = []
    for mult in multiplication:
        left_mult = number // mult
        addr_strs.append(str(left_mult))
        number -= left_mult * mult
    return '.'.join(addr_strs)


add_ips(0, (2**32)-1)
print(len(ip_ranges))
print('\n'.join(ip_ranges))

File: 1070/b.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Arrays [][2]uint64

func (a Arrays) Len() int           { return len(a) }
func (a Arrays) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a Arrays) Less(i, j int) bool { return a[i][0] < a[j][0] }

var whiteList, blackList Arrays
var ipRanges []string
var ranges map[uint64]string
var multipliers [4]uint64

func main() {
	multipliers = [4]uint64{16777216, 65536, 256, 1}
	ranges = map[uint64]string{
		1: "32", 2: "31", 4: "30", 8: "29", 16: "28",
		32: "27", 64: "26", 128: "25", 256: "24", 512: "23",
		1024: "22", 2048: "21", 4096: "20", 8192: "19",
		16384: "18", 32768: "17", 65536: "16", 131072: "15",
		262144: "14", 524288: "13", 1048576: "12", 2097152: "11",
		4194304: "10", 8388608: "9", 16777216: "8", 33554432: "7",
		67108864: "6", 134217728: "5", 268435456: "4",
		536870912: "3", 1073741824: "2", 2147483648: "1", 4294967296: "0",
	}
	in := bufio.NewReader(os.Stdin)
	n := ReadNumber(in)
	blackList = make(Arrays, 0)
	whiteList = make(Arrays, 0)
	for i := 0; i < n; i++ {
		line, _ := in.ReadString('\n')
		addr_str := strings.TrimSpace(line[:len(line)-1])
		isBlackList := true
		if addr_str[0] == '+' {
			isBlackList = false
		}
		var cleanAddr []string
		x := 32
		if strings.ContainsAny(addr_str, "/") {
			splittedList := strings.Split(addr_str[1:], "/")
			cleanAddr = strings.Split(splittedList[0], ".")
			x, _ = strconv.Atoi(splittedList[1])
		} else {
			cleanAddr = strings.Split(addr_str[1:], ".")
		}
		addrNum := uint64(0)
		for i := 0; i < 4; i++ {
			atoi, _ := strconv.Atoi(cleanAddr[i])
			addrNum = addrNum + multipliers[i]*uint64(atoi)
		}
		addressRange := math.Pow(2, float64(32-x))
		if isBlackList {
			blackList = append(blackList, [2]uint64{addrNum, addrNum + uint64(addressRange) - 1})
		} else {
			whiteList = append(whiteList, [2]uint64{addrNum, addrNum + uint64(addressRange) - 1})
		}
	}
	sort.Sort(blackList)
	sort.Sort(whiteList)
	ipRanges = make([]string, 0)
	addIPs(0, (2<<31)-1)
	fmt.Println(len(ipRanges))
	fmt.Println(strings.Join(ipRanges, "\n"))
}

func addIPs(left, right uint64) {
	if checkList(blackList, left, right) {
		if checkList(whiteList, left, right) {
			if left == right {
				fmt.Println("-1")
				os.Exit(0)
			}
			boundary := left + ((right - left) / 2)
			addIPs(left, boundary)
			addIPs(boundary+1, right)
		} else {
			ipRanges = append(ipRanges, fmt.Sprintf("%s/%s", numberToAddress(left), ranges[right-left+1]))
		}
	}
}

func checkList(listToCheck Arrays, left, right uint64) bool {
	if len(listToCheck) > 0 {
		el := listToCheck[findIndex(listToCheck, left)]
		if (left <= el[0] && el[0] <= right) || (left <= el[1] && el[1] <= right) || (el[0] <= left && left <= el[1]) || (el[0] <= right && right <= el[1]) {
			return true
		}
	}
	return false
}

func findIndex(listToCheck Arrays, index uint64) int {
	left := 0
	right := len(listToCheck) - 1
	for {
		if left == right {
			return left
		}
		mid := left + ((right - left) / 2)
		if listToCheck[mid][0] <= index && index <= listToCheck[mid][1] {
			return mid
		}
		if listToCheck[mid][0] > index {
			right = mid
		} else {
			left = mid + 1
		}
	}
}

func numberToAddress(number uint64) string {
	addrParts := [4]string{}
	for i := 0; i < 4; i++ {
		left_mult := number / multipliers[i]
		addrParts[i] = strconv.Itoa(int(left_mult))
		number = number - (left_mult * multipliers[i])
	}
	return strings.Join(addrParts[:], ".")
}

func ReadNumber(in *bufio.Reader) int {
	line, _ := in.ReadString('\n')
	num, _ := strconv.Atoi(strings.TrimSpace(line[:len(line)-1]))
	return num
}
